Replace region argument high bound behavior with more natural encapsulation

lydia-duncan opened the pull request:
As discussed in #20672, we realized that it made more sense for the region
argument to represent the whole portion of the file being covered, rather
than having the high bound be excluded from the region.  This PR adds a
flag per impacted function and a deprecation warning that triggers when
the flag is not set and an explicit high bound is used.

While here, update some documentation that had gotten out of date
(both w.r.t. the region argument and w.r.t. channels being split into
fileReader and fileWriter types)

Impacts:
- openreader
- file.reader
- file.writer
- file.lines
- channel.seek

Resolves #20672 
Resolves Cray/chapel-private#3858

To avoid generating the deprecation warning within the URL package
module, adds an argument to `file.readerHelper` and `file.writerHelper`
indicating that the call came from that source, so that the new behavior
is just used as part of the implementation.  Also adds an argument to
`file.readerHelper` in support of `openreader` so that a user can update
all their calls to `openreader` without needing to update any calls to
`file.reader` at the same time (due to `openreader`'s dependence on
`file.reader`).

Adds a test of all the new warnings and updates test files to set the
appropriate flag and adjust their specified boundaries as appropriate.

Passed a full paratest with futures and checked the built docs

jeremiah-corrado commented on modules/packages/URL.chpl:
```diff
@@ -147,8 +149,10 @@ module URL {
+    // TODO: change this back to f.writer when the fromOpenUrlWriter arg is
+    // removed
+    return f.writerHelper(kind=kind, locking=locking,
+                          region=start..end, fromOpenUrlWriter=true);
```
should this region argument also be `start..#end`?

jeremiah-corrado replied:
Maybe `openURLWriter` needs a test as well? I'm not seeing one in the test directory

lydia-duncan replied:
Oops, yes!

lydia-duncan replied:
There are tests of openUrlWriter in `test/library/packages/Curl`, they just require Curl is installed (which isn't true on all of our testing boxes).  So I suspect you just saved me from a nightly failure, but that we would have caught it then

jeremiah-corrado replied:
Oh that makes sense. Glad I saw that!

jeremiah-corrado commented on modules/standard/IO.chpl:
```diff
@@ -2620,6 +2624,20 @@ proc _channel.seek(region: range(?)) throws {
   }
 }
 
+deprecated "Currently the region argument high bound specifies the first location in the file that is not included.  This behavior is deprecated, please compile your program with `-suseNewSeekRegionBounds=true` to have the region argument specify the entire segment of the file covered, inclusive."
```
I might reword slightly: "Currently the region argument's high bound..."

jeremiah-corrado commented on modules/standard/IO.chpl:
```diff
@@ -2965,14 +3012,33 @@ proc file.readerHelper(param kind=iokind.dynamic, param locking=true,
+          fromOpenUrlReader ||
+          (!fromOpenReader && !fromOpenUrlReader
+           && useNewFileReaderRegionBounds)) {
+        ret = new fileReader(kind, locking, this, err, hints, region.low,
```
This expression feels complex enough that a comment explaining what it's checking might be helpful. Just in case someone has to modify this method before the deprecated behavior gets removed.

lydia-duncan replied:
I've pushed a comment, does it make sense to you?  I also realized that the `!fromOpenUrlReader` portion of it didn't really make sense, so I removed it

jeremiah-corrado replied:
Yeah that makes sense. Thanks!

lydia-duncan replied:
Thanks!  I'm running through a full paratest now just in case the conditional update broke anything

jeremiah-corrado commented on modules/standard/IO.chpl:
```diff
@@ -2829,7 +2849,17 @@ This function is equivalent to calling :proc:`open` and then
+  return openreaderHelper(path, kind, locking, region, hints);
+}
+
+deprecated "Currently the region argument high bound specifies the first location in the file that is not included.  This behavior is deprecated, please compile your program with `-suseNewOpenReaderRegionBounds=true` to have the region argument specify the entire segment of the file covered, inclusive."
```
same wording suggestion as above...

jeremiah-corrado reviewed (approved):
Looks good to me (I enjoyed your test text selections)!

Just one question on the openURLWriter implementation, and then a couple small suggestions...

lydia-duncan commented:
@jeremiah-corrado - your `channel.write` PR reminded me that I hadn't added documentation for these configs.  I've added it now and fixed some out-of-date information I noticed as well, would you mind taking another look?  Thanks

jeremiah-corrado reviewed (approved):
lgtm!
